File: tree/112-路径总和/112-路径总和.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */
 var hasPathSum = function(root, targetSum) {
    if(!root) return false
    // 设置 res 记录结果
    let res = false
    // dfs深度优先遍历函数，，进行递归调用---因为要计算路径总和因此必须将每个节点及其以上的节点和传入函数。
    const dfs = (n,s)=>{
        // 当递归到叶子节点时，判断是否与目标总和相同
        if(!n.left && !n.right && targetSum === s){
                res =  true ;
        }
        // 当左右子节点存在时--递归调用，并将当前路径和传递到函数中
        if(n.left) dfs(n.left,s+n.left.val)
        if(n.right) dfs(n.right,s+n.right.val)
    }

    dfs(root,root.val)
    return res
};

File: Python/112. 路径总和.py
# -*- codeing = utf-8 -*-
# @Time : 2023-8-18 22:27
# @Author : 刘永奇
# @File : 112. 路径总和.py
# @Software : PyCharm
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        # 如果为空二叉树直接返回False
        if root == None: return False
        # 当到达叶子节点时判断目标与当前节点值是否相等，如果相等返回True否则返回False
        if not root.left and not root.right:
            return targetSum == root.val
        # 递归调用，左右子树中是否有符合条件的
        return self.hasPathSum(root.left,targetSum - root.val) or self.hasPathSum(root.right,targetSum-root.val)